File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/system/SolidDataDirectory.java
package ch.bailu.aat_lib.preferences.system;

import java.util.ArrayList;

import ch.bailu.aat_lib.preferences.SolidFile;
import ch.bailu.aat_lib.resources.Res;
import ch.bailu.foc.FocFactory;

public class SolidDataDirectory extends SolidFile {

    private final SolidDataDirectoryDefault defaultDirectory;

    public SolidDataDirectory(SolidDataDirectoryDefault defaultDirectory, FocFactory focFactory) {
        super(defaultDirectory.getStorage(), SolidDataDirectory.class.getSimpleName(), focFactory);
        this.defaultDirectory = defaultDirectory;
    }

    @Override
    public String getLabel() {
        return Res.str().p_directory_data();
    }

    @Override
    public String getValueAsString() {
        String r = super.getValueAsString();


        if (getStorage().isDefaultString(r))
            return defaultDirectory.getValueAsString();

        return r;
    }

    @Override
    public boolean hasKey(String s) {
        return super.hasKey(s) || defaultDirectory.hasKey(s);
    }

    @Override
    public ArrayList<String> buildSelection(ArrayList<String> list) {
        return defaultDirectory.buildSelection(list);
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/system/SolidDataDirectory.kt
package ch.bailu.aat_lib.preferences.system

import ch.bailu.aat_lib.preferences.SolidFile
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.resources.Res
import ch.bailu.aat_lib.util.extensions.addUnique
import ch.bailu.foc.FocFactory

open class SolidDataDirectory(storageInterface: StorageInterface, focFactory: FocFactory)
    : SolidFile(storageInterface, SolidDataDirectory::class.java.simpleName, focFactory) {

    override fun getLabel(): String {
        return Res.str().p_directory_data()
    }

    override fun getValueAsString(): String {
        val r = super.getValueAsString()
        return if (getStorage().isDefaultString(r)) setDefaultValue() else r
    }

    fun buildSubDirectorySelection(result: ArrayList<String>, subDirectory: String): ArrayList<String> {
        result.addUnique("${getValueAsString()}/${subDirectory}")
        buildSelection().forEach {
            result.addUnique("${it}/${subDirectory}")
        }
        return result
    }
}
